Bugs fixes and enhancements

gcopeland opened the pull request:
My master has changes which include a write() bug fix and dynamic timeout adjustment, multi SPI device compatibility, additional printDetails() debugging output (nice improvements BTW), some comment clean ups, a US/internationally friendly default channel, and some minor changes to pingpair for my own local testing. The pingpair changes can be ignored.

The bug fix on write() now monitors the proper radio register. Also included in this code is deadlock timeout logic which dynamically adjusts based on current ARD/ARC hardware configuration.

IMOHO, the bugfix and dynamic timeout adjustment alone is worth the pull.

Cheers,

gcopeland committed:
Broken out SPI bus configuratino to its own method. SPI access is not
prefixed with this new private method to ensure proper operation when
multiple SPI devices are in use.

gcopeland committed:
Hopefully fix a radio write bug.

gcopeland committed:
Changed SIP bus speed to use a divider of 4 for enhanced
compatibility.

gcopeland committed:
Merge branch 'master' of https://github.com/maniacbug/RF24

gcopeland committed:
Merge clean up.

gcopeland committed:
Added some additional debug options (commented out) for testing. More
merge cleanup (printDetails).

gcopeland committed:
Yet more merge cleanup. Changes default channel to be
US/Internationally legal while ensuring no spectrum bleed.

maniacbug commented:
Thanks, I look forward to checking it out.  Have you merged my master in prior to this?

-----Original Message-----
From: gcopeland
Sent: Wednesday, August 17, 2011 10:47 AM
To: maniacbug@ymail.com
Subject: [RF24] Bugs fixes and enhancements (#2)

My master has changes which include a write() bug fix and dynamic timeout adjustment, multi SPI device compatibility, additional printDetails() debugging output (nice improvements BTW), some comment clean ups, a US/internationally friendly default channel, and some minor changes to pingpair for my own local testing. The pingpair changes can be ignored.

The bug fix on write() now monitors the proper radio register. Also included in this code is deadlock timeout logic which dynamically adjusts based on current ARD/ARC hardware configuration.

IMOHO, the bugfix and dynamic timeout adjustment alone is worth the pull.

Cheers,

## 

Reply to this email directly or view it on GitHub:
https://github.com/maniacbug/RF24/pull/2

maniacbug commented:
Ok, I see now that you did merge the master in.  Yay.  That should make it a little easier.
1. Calculated write() timeouts.  Looks like a good enhancement.  Will certainly fold this in.
2. SPI speed.  As long as it passes the tests, its fine w/ me.
3. Default channel.  Sounds reasonable.  Could you write up something for the docs to explain why it's the best choice?
4. printf_P PA Power. Looks good.  I see you got the hang of how I want to do these.  :)  You can also see that I did want this info, just wanted it printed nicely.

Regarding a bug in 'write()'.  I do not see a bug in write() in the master branch of my repo.  "status = read_register(OBSERVE_TX,&observe_tx,1);" is a perfectly legitimate way of reading the status.  It's that way so that in SERIAL_DEBUG mode I can print out the value of observe_tx, otherwise it's ignored.

There definitely was a bug in the version you had, but you had a pretty old drop.  I had fixed the bug in there already.

For the SPI stuff, it's not clear what is being done here.  What is the user benefit of b76284700d86adb9c928 ?

gcopeland committed:
Performance optimizations. Radio no longer automatically powers down.
Radio is now allowed to enter standby. This allows for the removal of
needless delays within the driver. Additional timings optimizations.
If applications desire additional power savings, they should manually
power down the radio as needed.

Multicast write support added.

::write timeout monitoring logic changed to emcompass worst timeouts
for retries/timeouts. Should never trigger anyways...but better safe.

Fixes write retries timeout bug.

Proper hardware delay added to powerUp. Was previously part of
startWrite; which needlessly added overhead to every write/startWrite.

Allows for query of hardware channel.

PA documentation fixes.

gcopeland committed:
Previous commit message applies here.

gcopeland committed:
Fixed write hardware timeout monitoring. Timeout is now dynamically
calculated based on current hardware configuration. There is room
for optimization via caching.

gcopeland committed:
Fixed spelling.

gcopeland committed:
Changed timeout calculation from ms to us as usual math resulted in
loss of resolution and the possibility of a premature timeout. The
write monitoring loop now uses us rather than ms to be more compatible
with the us result returned from getMaxTimeout.

gcopeland committed:
Normalized write and startWrite interfaces for multicast. Now both take
a bool for multicasting. Documentation updates.

gcopeland committed:
Updated documentation for multicast warnings.

gcopeland committed:
More doc updates.

gcopeland committed:
Minor cleanup. Documentation change for multicast.

gcopeland committed:
Docs.

gcopeland committed:
Example changes. Pingpair now reports a more accurate round drip without
artificial delay. Examples use more options. Added a new multicast
examples based on pingpair_dyn; pingpair_multi_dyn.

gcopeland committed:
Removed redundant headers as its already included, and required, by
RF24.h. Fixed some radio power up/down issues in code which relied
on side effect behavior of radio powering down after every TX.
Applications now take responsibility for power management.

gcopeland committed:
Removed redundant arduino < 100 header include.

gcopeland committed:
More cleanup.

gcopeland committed:
Cleanup, fork documentations to README.md, and use of dynamic
timeout in tests.

gcopeland committed:
Added closeReadingPipe method and extended keywords with common
method calls.

gcopeland committed:
Removed redundant rx/tx buffer flushes. If an application needs
these flushed, it should do it. Otherwise, properly behaving
applications should need this. Also, by not flushing it opens
the door for additional retransmission features down the road.

gcopeland committed:
Updated scanner to be a bit more sensitive as well as much easier
to read.

gcopeland committed:
Bug fix. Added back 1MBPS RPD check.

gcopeland committed:
Discovered old debugging SPI bus clock setting. This horribly
slows things down. Its now where it needs to be. Dang.

farconada commented:
What about this PR? I'm considering gcopeland version but I think that It should be merged

karanlyons commented:
This PR would be great to merge in, especially the multi SPI support (the commit you called out, @maniacbug).

Currently your RF24 library assumes it’s the only thing using SPI, and that the NRF is the only slaved device. If you have multiple devices that you want to communicate with, your library will likely break because it doesn’t properly reset the SPI configuration before attempting to communicate with the radio (so it’ll get whatever configuration was last set, which’ll be for another device, and may not match the NRF’s spec).

FrankOcean11 commented:
im having issues tx values from imu to map calls i made for servo use, i notice in terminal i got some values being produce by the imu in master but in the terminal window for slave i have a servo and it only moves IF i restart the master terminal and only then it will tx to the rx terminal only one time  so this mens i have to keep resetting the window so that i can tx imu value from a array to the slave and by the time that happens im sending old values whos data you cant match to the tx terminal, any clue why or is this fix you made to the write() is the reason why ? 
![capture](https://cloud.githubusercontent.com/assets/10504709/10552544/97d6c134-7422-11e5-80f4-55f09c8cbdc2.JPG)
